### PR TITLE
Restricted calls simplification

### PR DESCRIPTION
**Pull request by istankovic**
We now handle everything inside a single contract, RestrictedCalls.
The IMessenger interface got a new function, callAllowed which is
supposed to say whether the call is allowed or not. This completely
pushes call checking to messengers, which are chain-specific.
For same-chain calls, the messengers are not consulted since we can
check directly whether the caller is allowed to make the call.

### CODE REVIEW

File: contracts/contracts/RestrictedCalls.sol [hunk 1]
```diff
@@ -30,12 +53,29 @@ contract RestrictedCalls is Ownable {
...
+    ///     restricted(otherChainId)    // expecting calls from another chain
     ///
-    modifier restricted(uint256 chainId, address caller) {
-        bytes32 key = keccak256(abi.encodePacked(chainId, caller));
+    modifier restricted(uint256 callerChainId) {
+        address caller = callers[callerChainId][0];
 
-        require(callers[key], "RestrictedCalls: unknown caller");
+        if (callerChainId == block.chainid) {
+            require(msg.sender == caller, "RestrictedCalls: unknown caller");
```
Comment (fredo) [inline contracts/contracts/RestrictedCalls.sol:63]: can'T we unify the revert message with the one below? "RestrictedCalls: call disallowed"

File: contracts/contracts/RestrictedCalls.sol [hunk 2]
```diff
@@ -2,25 +2,48 @@
...
 /// to a single address. This allows for a trusted call chain,
 /// as described in :ref:`contracts' architecture <contracts-architecture>`.
-///
-/// .. seealso:: :sol:contract:`CrossDomainRestrictedCalls`
 contract RestrictedCalls is Ownable {
-    /// The set of callers, in the form of hashed pairs ``(chainId, address)``.
-    mapping(bytes32 => bool) public callers;
+    /// Maps caller chain IDs to tuples [caller, messenger].
+    ///
+    /// For same-chain calls, the messenger address is 0.
```
Comment (fredo) [inline contracts/contracts/RestrictedCalls.sol:13]: nitpick: a better (more common) representaion forthe zero address would be 0x0 or address(0)

File: contracts/contracts/RestrictedCalls.sol [hunk 3]
```diff
@@ -2,25 +2,48 @@
...
+        require(caller != address(0), "RestrictedCalls: caller cannot be 0");
+        require(
+            callers[callerChainId][0] == address(0),
+            "RestrictedCalls: caller already exists"
+        );
+        callers[callerChainId] = [caller, messenger];
+    }
 
-    /// Add a caller for the given chain ID.
+    /// Allow calls from a contract on the same chain.
```
Comment (fredo) [inline contracts/contracts/RestrictedCalls.sol:29]: `Allow calls from an address`

File: contracts/contracts/RestrictedCalls.sol [hunk 4]
```diff
@@ -2,25 +2,48 @@
...
     /// @param caller The caller.
-    function addCaller(uint256 chainId, address caller) external onlyOwner {
-        bytes32 key = keccak256(abi.encodePacked(chainId, caller));
+    function addCaller(address caller) external onlyOwner {
+        _addCaller(block.chainid, caller, address(0));
+    }
 
-        require(!callers[key], "RestrictedCalls: caller already exists");
-        callers[key] = true;
+    /// Allow calls from a contract on another chain.
```
Comment (fredo) [inline contracts/contracts/RestrictedCalls.sol:36]: Allow calls from an address

File: contracts/contracts/chains/arbitrum/ArbitrumMessengers.sol [hunk 5]
```diff
@@ -92,28 +93,24 @@ contract ArbitrumL1Messenger is IMessenger, RestrictedCalls {
...
-    function nativeMessenger() external view returns (address) {
+    function callAllowed(
+        uint256,
+        address caller,
+        address courier
+    ) external view returns (bool) {
         // Arbitrum sets `msg.sender` on calls coming from L1 to be the aliased
-        // address of the contract that sent the message so we need to return that
-        // aliased address here. In our case, that is the L2 alias of our ArbitrumL1Messenger.
+        // address of the contract that sent the message so we need to check
```
Comment (fredo) [inline contracts/contracts/chains/arbitrum/ArbitrumMessengers.sol:102]: ...of the address...

File: contracts/contracts/RestrictedCalls.sol (hunk 1 above)
Comment (istankovic) [reply contracts/contracts/RestrictedCalls.sol:63]: Good point. Done.

**Comment by istankovic** [comment]
The branch has been rebased on top of latest `main` and all review comments made so far have been addressed.

**Comment by istankovic** [comment]
I just confirmed the local Arbitrum L1 resolution test (local Arbitrum -> local geth -> local Arbitrum) works with changes in this PR so from my side this is cleared to land.

**Review by fredo — COMMENTED** [review]
Looks awesome! I like the new design. 

One question, which is left unanswered for me. Why do we have the callerChainId in the respective `callAllowed()` interface? 
We are not using it anywhere and as far as I understand it, we would only need it when one specific L1Messenger would be responsible for multiple chains. This is something which is neither intended in our design nor possible with the current messenger interface. 
What did I miss?

**Comment by istankovic** [comment]
> One question, which is left unanswered for me. Why do we have the callerChainId in the respective `callAllowed()` interface?
[...]

Good catch. Originally I went with it because I thought it might be required in messenger implementations, but as it turns out, it is not actually required. I guess the other reason was that I thought it nice to give the messengers complete caller information `(callerChainId, caller)`, as opposed to just `caller`.

Shall I just remove that parameter from the interface?

**Comment by fredo** [comment]
> > One question, which is left unanswered for me. Why do we have the callerChainId in the respective `callAllowed()` interface?
> > [...]
> 
> Good catch. Originally I went with it because I thought it might be required in messenger implementations, but as it turns out, it is not actually required. I guess the other reason was that I thought it nice to give the messengers complete caller information `(callerChainId, caller)`, as opposed to just `caller`.
> 
> Shall I just remove that parameter from the interface?

I understand. In this case, I would prefer to remove it. As I mentioned, it would require a restructuring of the messenger itself to make him use of the chain id. So I'd prefer to not carry it along the way and potentially add it once we ever need it.

**Comment by istankovic** [comment]
Removed the `callerChainId` parameter from the interface.

**Comment by istankovic** [comment]
Here's the final gas profile:

```
FillManager <Contract>
   ├─ constructor            -  avg:  833666  avg (confirmed):  833666  low:  833661  high:  833673
   ├─ fillRequest            -  avg:   75832  avg (confirmed):  101368  low:   23803  high:  136897
   ├─ addAllowedLp           -  avg:   44201  avg (confirmed):   44626  low:   22744  high:   44629
   ├─ setResolver            -  avg:   43409  avg (confirmed):   44302  low:   22699  high:   44305
   ├─ invalidateFill         -  avg:   38586  avg (confirmed):   48504  low:   23073  high:   82132
   └─ removeAllowedLp        -  avg:   14822  avg (confirmed):   14822  low:   14819  high:   14825
MintableToken <Contract>
   ├─ constructor            -  avg:  679653  avg (confirmed):  679653  low:  679653  high:  679653
   ├─ mint                   -  avg:   50694  avg (confirmed):   50694  low:   50676  high:   50784
   └─ approve                -  avg:   42867  avg (confirmed):   42867  low:   24884  high:   44468
RequestManager <Contract>
   ├─ constructor            -  avg: 2706240  avg (confirmed): 2706240  low: 2706240  high: 2706240
   ├─ createRequest          -  avg:  161203  avg (confirmed):  178014  low:   24194  high:  225941
   ├─ claimRequest           -  avg:  144266  avg (confirmed):  164669  low:   22675  high:  168968
   ├─ addCaller              -  avg:   66470  avg (confirmed):   66470  low:   66469  high:   66481
   ├─ challengeClaim         -  avg:   62340  avg (confirmed):   82756  low:   24292  high:  109716
   ├─ withdraw               -  avg:   58736  avg (confirmed):   88658  low:   22509  high:  142283
   ├─ withdrawExpiredRequest -  avg:   54279  avg (confirmed):   69090  low:   24658  high:   69090
   ├─ resolveRequest         -  avg:   52854  avg (confirmed):   62540  low:   23361  high:   67939
   ├─ addAllowedLp           -  avg:   44166  avg (confirmed):   44446  low:   22833  high:   44718
   ├─ setFinalityPeriod      -  avg:   43976  avg (confirmed):   43976  low:   43976  high:   43976
   ├─ updateFeeData          -  avg:   38056  avg (confirmed):   38282  low:   16754  high:   63604
   ├─ invalidateFill         -  avg:   37939  avg (confirmed):   47257  low:   23739  high:   47261
   ├─ deprecateContract      -  avg:   29699  avg (confirmed):   43788  low:   22239  high:   43788
   ├─ updateTransferLimit    -  avg:   26248  avg (confirmed):   27473  low:   22575  high:   27477
   ├─ withdrawProtocolFees   -  avg:   23670  avg (confirmed):   23203  low:   23203  high:   24133
   └─ removeAllowedLp        -  avg:   14815  avg (confirmed):   14815  low:   14815  high:   14815
Resolver <Contract>
   ├─ constructor            -  avg:  678524  avg (confirmed):  678524  low:  678524  high:  678524
   ├─ addCaller              -  avg:   66416  avg (confirmed):   66416  low:   66415  high:   66427
   ├─ addRequestManager      -  avg:   65139  avg (confirmed):   65139  low:   65122  high:   65146
   └─ resolve                -  avg:   23388  avg (confirmed):       0  low:   23268  high:   23508
TestL1Messenger <Contract>
   ├─ constructor            -  avg:  532868  avg (confirmed):  532868  low:  532868  high:  532868
   ├─ addCaller              -  avg:   46577  avg (confirmed):   46577  low:   46568  high:   46580
   ├─ setLastSender          -  avg:   42658  avg (confirmed):   42658  low:   42658  high:   42658
   ├─ setForwardState        -  avg:   24935  avg (confirmed):   24935  low:   23239  high:   42451
   └─ sendMessage            -  avg:   23400  avg (confirmed):       0  low:   23400  high:   23400
TestL2Messenger <Contract>
   ├─ constructor            -  avg:  532868  avg (confirmed):  532868  low:  532868  high:  532868
   ├─ addCaller              -  avg:   46576  avg (confirmed):   46576  low:   46568  high:   46580
   ├─ setForwardState        -  avg:   24935  avg (confirmed):   24935  low:   23239  high:   42451
   └─ sendMessage            -  avg:   23400  avg (confirmed):       0  low:   23400  high:   23400
```

**Review by fredo — APPROVED** [review]

**Pull request merged**